packages/contracts/contracts/L1/tss/TssStakingSlashing.sol

```diff
@@ -0,0 +1,373 @@
+    }
+
+    /**
+     * @notice slash the staker and distribute rewards to voters
+     * @param message the message about the slash infos
+     */
+    function slash(SlashMsg memory message) internal {
+        // slashing params check
+        for (uint256 i = 0; i < 2; i++) {
+            require(slashAmount[i] > 0, "have not set the slash amount");
+            require(exIncome[i] > 0, "have not set the extra income amount");
+        }
+        bytes memory jailNodePubKey = deposits[message.jailNode].pubKey;
+        if (message.slashType == SlashType.uptime) {
+            // jail and transfer deposits
+            ITssGroupManager(tssGroupContract).memberJail(jailNodePubKey);
+            transformDeposit(message.jailNode, 0, message.tssNodes);
+        } else if (message.slashType == SlashType.animus) {
+            // remove the member and transfer deposits
+            ITssGroupManager(tssGroupContract).removeMember(jailNodePubKey);
+            transformDeposit(message.jailNode, 1, message.tssNodes);
+        } else {
+            require(false, "err type for slashing");
+        }
+    }
+
+    /**
+     * @notice distribute rewards to voters
+     * @param deduction address of the punished
+     * @param slashType the type to punished
+     * @param tssNodes participants other than the initiator
+     */
+    function transformDeposit(
+        address deduction,
+        uint256 slashType,
+        address[] memory tssNodes
+    ) internal {
+        uint256 deductedAmount;
+        uint256 totalTransfer;
+        uint256 extraAmount;
+        uint256 remainder;
+        uint256 gain;
+
+        if (deposits[deduction].amount > slashAmount[slashType]) {
+            // deposit > slashAmount, deduct slashAmount then
+            // distribute additional tokens for the sender
+            deductedAmount = slashAmount[slashType];
+            deposits[deduction].amount -= slashAmount[slashType];
+            extraAmount = slashAmount[slashType] - exIncome[slashType];
+        } else if (deposits[deduction].amount > exIncome[slashType]) {
+            // exIncome < deposit <= slashAmount, deduct all token then
+            // distribute additional tokens for the sender
+            deductedAmount = deposits[deduction].amount;
+            deposits[deduction].amount = 0;
+            extraAmount = deductedAmount - exIncome[slashType];
+        } else if (deposits[deduction].amount > 0) {
+            // 0 < deposit <= exIncome, deduct all token
+            deductedAmount = deposits[deduction].amount;
+            deposits[deduction].amount = 0;
+            extraAmount = deductedAmount;
+        } else {
+            require(false, "panic , invalid type");
+        }
+
+        // transfer sender the remainder as an additional bonus
+        remainder = extraAmount % tssNodes.length;
+        gain = (extraAmount - remainder) / tssNodes.length;
+
+        deposits[msg.sender].amount += exIncome[slashType] + remainder;
```

Comment:
in this case of "0 < deposit <= exIncome", we can't allocate any reward to sender. since according to your code, all the deposits are distributed to tssNodes, none are left for sender

Reply:
I read the code again and found that for the case of 0 < deposit <= exIncome, the sender will distribute more tokens

Reply:
https://github.com/bitnetworkio/bitnetwork/pull/68/commits/ba4de383aa34154f6cb58462b37c4ca886f3315b#diff-eff8d07bbf84d1629b66497fbdb9e9353e16f5b6bcaaff81b53744933ed99d6eL287